pep-0544.txt

```diff
@@ -439,98 +448,123 @@ but a static type checker will handle them specially. Subclassing a protocol
 
   from typing import Sized, Protocol
 
-  class SizedAndCloseable(Sized, Protocol):
+  class SizedAndClosable(Sized, Protocol):
       def close(self) -> None:
           ...
 
-Now the protocol ``SizedAndCloseable`` is a protocol with two methods,
+Now the protocol ``SizedAndClosable`` is a protocol with two methods,
 ``__len__`` and ``close``. If one omits ``Protocol`` in the base class list,
 this would be a regular (non-protocol) class that must implement ``Sized``.
-If ``Protocol`` is included in the base class list, all the other base classes
-must be protocols. A protocol can't extend a regular class.
-
-Alternatively, one can implement ``SizedAndCloseable`` like this, assuming
-the existence of ``SupportsClose`` from the example in `definition`_ section::
+Alternatively, one can implement ``SizedAndClosable`` protocol by merging
+the ``SupportsClose`` protocol from the example in `definition`_ section
+with ``typing.Sized``::
 
   from typing import Sized
 
-  class SupportsClose(...): ...  # Like above
+  class SupportsClose(Protocol):
+      def close(self) -> None:
+          ...
 
-  class SizedAndCloseable(Sized, SupportsClose, Protocol):
+  class SizedAndClosable(Sized, SupportsClose, Protocol):
       pass
 
 The two definitions of ``SizedAndClosable`` are equivalent.
 Subclass relationships between protocols are not meaningful when
 considering subtyping, since structural compatibility is
 the criterion, not the MRO.
 
-Note that rules around explicit subclassing are different from regular ABCs,
-where abstractness is simply defined by having at least one abstract method
-being unimplemented. Protocol classes must be marked *explicitly*.
+If ``Protocol`` is included in the base class list, all the other base classes
+must be protocols. A protocol can't extend a regular class, see `rejected`_
+ideas for rationale. Note that rules around explicit subclassing are different
+from regular ABCs, where abstractness is simply defined by having at least one
+abstract method being unimplemented. Protocol classes must be marked
+*explicitly*.
 
 
-Generic and recursive protocols
--------------------------------
+Generic protocols
+-----------------
 
 Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
 and ``Iterator`` are generic protocols. They are defined similar to normal
 non-protocol generic types::
 
-  T = TypeVar('T', covariant=True)
-
   class Iterable(Protocol[T]):
       @abstractmethod
       def __iter__(self) -> Iterator[T]:
           ...
 
-Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
-``Protocol, Generic[T, S, ...]``.
+``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``. Declaring variance is not necessary for
+protocol classes, since it can be inferred from a protocol definition.
+Example::
+
+  class Box(Protocol[T]):
+      content: T
+
+  x: Box[float]
+  y: Box[int]
+  x = y  # This is OK due to inferred covariance of the protocol 'Box'.
```

Comment:
Since the `content` attribute is mutable, `Box` should be invariant? And for consistency with normal classes, I'd prefer variance to be explicit. Especially if a generic protocol inherits from other protocols, it can be unclear whether a particular type variable should be covariant or invariant, so making this explicit would make code easier to understand. Users are often confused by variance, so it's likely that they often can't predict what the inferred variance would be. And what if I'd like to explicitly make something invariant for some reason, even if inference thinks it can be covariant?